doc: tfm: add overview of PSA Certified API (#20863)
nrfconnect/sdk-nrf: doc/nrf/security/tfm/psa_certified_api_overview.rst

.. _ug_psa_certified_api_overview:

PSA Certified Security Framework overview
#########################################

.. contents::
   :local:
   :depth: 2

IoT devices connect to the internet and exchange information with other devices, which makes them targets for hackers who exploit security vulnerabilities.
Protecting data integrity, confidentiality, and availability on these devices has therefore become a priority.

The IoT industry has for a long time lacked a common security language and standardized processes.
This fragmentation has led to inconsistent and often inadequate security across devices.

> Product developers often view security as complex and expensive, potentially increasing time-to-market and harming business outcomes.

Reviewer: Line 10-15:
Is this too "marketing" for documentation? Does it serve any purpose here?
I think maybe these lines can be removed without losing any information

The `PSA Certified IoT Security Framework`_ offers a standardized approach to create secure IoT devices, covering everything from analysis to certification.
It also provides free `resources <PSA Certified development resources_>`_ and `training material <PSA Certified foundational training course_>`_ to educate developers about threat modeling, certification, and other security topics.

The framework divides security design into a simple four-step process to guide product developers into implementing the appropriate level of security in their IoT devices:

* Analyze the threats that have the potential to compromise your device and generate a set of security requirements based on these risks.
* Architect the right level of security for your product by using unique security requirements to identify and select components and specifications.
* Implement the trusted components and firmware, making use of high-level APIs to build-in security and create an interface to the hardware Root of Trust (RoT).

> * Certify device, platform, or silicon by following independent security evaluation.

Author: They are not on the source page.

.. figure:: images/psa_certified_api_general.png
   :alt: PSA Crypto API usage
   :align: center

   PSA Crypto API usage

The `PSA Certified APIs`_ are a set of high-level standardized software interfaces that aim to enhance the security of IoT devices.
They are designed to support a range of different use cases and are organized around the following groups:

* `PSA Certified Crypto API`_
* `PSA Certified Secure Storage API`_
* `PSA Certified Attestation API`_
* `PSA Certified Firmware Update API`_

> Using the PSA Certified APIs has numerous benefits:

Reviewer 2: ```suggestion
Using the PSA Certified APIs has the following benefits:
```

* Enhanced security - PSA Certified APIs enable devices to meet industry-standard security requirements and enhance the overall trustworthiness of IoT devices.

> * Implementation agnostic - By using PSA Certified APIs, developers don't need to be concerned about the underlying hardware and software implementation.

Reviewer 2: ```suggestion
* Implementation agnostic - When using PSA Certified APIs, developers do not need to be concerned about the underlying hardware and software implementation.
```

* Reduce time-to-market - Using an API standard can accelerate development time and reduce costs associated with developing, testing, and certifying custom solutions.
* Flexible and scalable - The various use cases supported ensure that the PSA Certified APIs can be used across multiple devices, from very simple ones to more complex systems.

> * Future-proof - PSA Certified APIs are designed to be updated over time as security threats evolve, ensuring devices remain secure throughout their lifecycle.

Reviewer 2: ```suggestion
* Future-proof - PSA Certified APIs are designed to be updated over time as security threats evolve, ensuring that devices remain secure throughout their lifecycle.
```

.. _ug_psa_certified_api_overview_crypto:

PSA Crypto API
**************

This section summarizes some of the features of the PSA Crypto API and how they can be specific to the implementation in the |NCS|.
For a full explanation, see the `PSA Certified Crypto API`_ official documentation.
See the `psa/crypto.h`_ file for versioning.

Among the advantages of the PSA Crypto API are the following:

* The PSA Crypto API is a single API for all cryptographic drivers, which means that you can use the same functions for the nRF52840, nRF5340, nRF54L15, and nRF9160 devices, as well as future ones.
  The API will work for applications with and without Trusted Firmware-M.

* The PSA Crypto API will automatically select cryptographic libraries based on project configurations.
  This way, the codebase for cryptography can easily be reused across multiple projects.

> The PSA Crypto API is designed to be safe and easy to use, lowering developers' possibility of introducing vulnerabilities into their end devices.

Reviewer 2: ```suggestion
The PSA Crypto API is designed to be safe, lowering developers' possibility of introducing vulnerabilities into their end devices.
```
Too much marketing jargon to my taste.

> * The PSA Crypto API is a single API for all cryptographic drivers, which means that you can use the same functions for the nRF52840, nRF5340 and nRF9160 devices, as well as future ones.

Reviewer: Also mention 54L15

See `Keystore Interface`_ for an overview.

PSA Crypto API in the |NCS|
===========================

The Crypto API is used to request cryptographic operations in the |NCS|.

> Supported operations include hashing, encryption and decryption, authenticated encryption, and signature generation and verification.

Reviewer 2: ```suggestion
Supported operations include the following:

* Hashing
* Encryption and decryption
* Authenticated encryption
* Signature generation and verification

See :ref:`crypto_samples` for usage examples.
See :ref:`nrf_security_drivers` for a list of supported functionalities.

> The Crypto API will use different libraries depending on hardware capabilities and user configuration:

Reviewer 2: ```suggestion
The Crypto API uses different libraries depending on hardware capabilities and user configuration:
```

> PSA Crypto API in the |NCS|

Reviewer: Somewhere in this chapter, we should mention that keys from the PSA Crypto API are typically stored in ITS or the KMU.

* cc3xx libraries - These provide support for the `CryptoCell 310 <nRF9160 CRYPTOCELL - Arm TrustZone CryptoCell 310_>`_ and `CryptoCell 312 <nRF5340 CRYPTOCELL — Arm TrustZone CryptoCell 312_>`_ hardware peripherals.

> * nrf_oberon - :ref:`nrf_oberon <nrfxlib:nrf_oberon_readme>` is a highly-optimized software library for cryptographic algorithms created by Oberon Microsystems and licensed to Nordic Semiconductor for redistribution.

Reviewer 2: ```suggestion
* nrf_oberon - :ref:`nrf_oberon <nrfxlib:nrf_oberon_readme>` is an optimized software library for cryptographic algorithms created by Oberon Microsystems and licensed to Nordic Semiconductor for redistribution.
```

See :ref:`crypto_samples` for usage examples.
See :ref:`nrf_security_drivers` for a list of supported functionalities.

The Crypto API will use different libraries depending on hardware capabilities and user configuration:

> * cc3xx libraries - These provide support for the `CryptoCell 310 <nRF9160 CRYPTOCELL - Arm TrustZone CryptoCell 310_>`_ and `CryptoCell 312 <nRF5340 CRYPTOCELL — Arm TrustZone CryptoCell 312_>`_ hardware peripherals.

Reviewer: Now that we got the nRF54L15 which uses CRACEN, we should make add a third option both here and in the figure below

> For a specific cryptographic operation, the PSA Crypto API will use the library :ref:`configured <nrf_security_drivers>` for the operation.

Reviewer 2: ```suggestion
For specific cryptographic operations, the PSA Crypto API uses the library :ref:`configured <nrf_security_drivers>` for them.
```

> If multiple libraries have been enabled, cc3xx will be prioritized, given that the CryptoCell supports the cryptographic operation.

Reviewer 2: ```suggestion
If multiple libraries have been enabled, cc3xx is prioritized, given that the CryptoCell supports the cryptographic operation.
```

+=======================+==================================+===================+
| Persistent Storage    | Yes                              | Yes               |
+-----------------------+----------------------------------+-------------------+
| Root of Trust (RoT)   | Platform RoT Service             | Application RoT   |
|                       |                                  | Service           |
+-----------------------+----------------------------------+-------------------+
| Available in the SPE  | Yes                              | Yes               |
+-----------------------+----------------------------------+-------------------+
| Available in the NSPE | No (by default - see note below) | Yes               |
+-----------------------+----------------------------------+-------------------+
| Encryption            | No (see note below)              | Yes               |
+-----------------------+----------------------------------+-------------------+
| Integrity Check       | No (see note below)              | Yes               |
+-----------------------+----------------------------------+-------------------+

.. note::

>    * The PSA Crypto API will be invoked by other APIs, which are available from the NSPE.

Reviewer 2: ```suggestion
   * The PSA Crypto API is invoked by other APIs that are available from the NSPE.
```

     For example the PSA Protected Storage API or the PSA Crypto API ``import()`` and ``generate()``.

>    * Internal Trusted Storage has experimental support for encryption with :kconfig:option:`CONFIG_PSA_ITS_ENCRYPTED`.

Reviewer 2: ```suggestion
   * Internal Trusted Storage has experimental support for encryption with the :kconfig:option:`CONFIG_PSA_ITS_ENCRYPTED` Kconfig option set.
```

PSA Secure Storage API in Nordic hardware and the |NCS|
=======================================================

> SoCs from Nordic Semiconductor that come with a :term:`System Protection Unit (SPU)` hardware peripheral (such as `nRF9160 <nRF9160 System Protection Unit_>`_) have the functionality for `flash access control <nRF9160 flash access control_>`_, making it possible to configure flash sections as secure.

Reviewer: It is no longer only the SPU that does this. Either also mention how this is done for the nRF54L15, or reword this part to be more generic

> SoCs from Nordic Semiconductor that come with dedicated hardware components for security (such as `nRF9160's System Protection Unit <nRF9160 System Protection Unit_>`_ or :ref:`nRF54L Series' security components <nRF54L15 Security_>`_) have the functionality for `flash <nRF9160 flash access control_>`_ (on nRF9160) or `feature <nRF54L15 feature access control_>`_ access control, making it possible to configure different features as secure.

Reviewer 2: ```suggestion
SoCs from Nordic Semiconductor that come with dedicated hardware components for security (such as `nRF9160's System Protection Unit <nRF9160 System Protection Unit_>`_ or `nRF54L Series' security components <nRF54L15 Security_>`_) have the functionality for `flash <nRF9160 flash access control_>`_ (on nRF9160) or `feature <nRF54L15 feature access control_>`_ access control, making it possible to configure different features as secure.
```

Internal Trusted Storage and Protected Storage will save data to the sections of secure flash.
This way, the NSPE can not directly access data saved by the Internal Trusted Storage API or the Protected Storage API.

> Internal Trusted Storage is only available internally.

Reviewer: ```suggestion
Internal Trusted Storage is by default only available from the SPE.
```

> Internal Trusted Storage and Protected Storage will save data to the sections of secure flash.

Reviewer 2: ```suggestion
Internal Trusted Storage and Protected Storage saves data to the sections of secure flash.
```

> For Nordic SoCs without an SPU, the PSA Secure Storage API will save data in regular flash storage.

Reviewer: This sentence should not mention the SPU, but be more generic. 
Also, it should link to the Trusted Storage docs, as those are used for devices that does not have HW to secure the storage.

> This way, the NSPE can not directly access data saved by the Internal Trusted Storage API or the Protected Storage API.

Reviewer 2: ```suggestion
This way, the NSPE cannot directly access data saved by the Internal Trusted Storage API or the Protected Storage API.
```

> To store sensitive user information in an external flash, it is recommended to use the Protected Storage API.

Reviewer: Can we double check if Protected Storage actually supports External flash?

> For Nordic SoCs without an SPU, the PSA Secure Storage API will save data in regular flash storage.

Reviewer 2: ```suggestion
For Nordic SoCs without an SPU, the PSA Secure Storage API saves data in the regular flash storage.
```

> This encrypts the data, protecting it in case the external flash is lost.

Reviewer 2: ```suggestion
This encrypts the data, protecting it if the external flash is lost.
```

Author: line removed

The Protected Storage API also performs integrity checks on the data, so that you know if the read data can still be trusted.

.. _ug_psa_certified_api_overview_attestation:

PSA Attestation API
*******************

This section summarizes some of the features of the PSA Attestation API.
For a full explanation, see the `PSA Certified Attestation API`_ official documentation.
See `psa/initial_attestation.h`_ for versioning.

See also the official `Device Attestation and Entity Attestation Tokens Explained`_ blog post from PSA Certified, which gives an overview of attestation.
This blog post explains well when to use the PSA Attestation API:

    Cloud service providers need to make informed judgements on end devices to ensure the data they are providing can be trusted.
    EAT has the capabilities to provide this source of trust, using a cryptographically signed piece of data containing claims that are generated in the device RoT.

>     There are many ways it can be useful, but most importantly it can be read by the relying party.

Reviewer 2: ```suggestion
    It can be useful in many ways, but most importantly it can be read by the relying party.
```

Author: Quote, no edit applied.

    The relying party can verify the claims made by the device such as:

    * The unique identity of the device
    * Installed software on the device and its integrity status
    * Security assurance and certification status
    * Manufacturer of the device hardware

    Using this information, the relying party can make informed decisions such as whether the device is legitimate and should be onboarded, or what services should be enabled based on its security credentials.

PSA Attestation API in the |NCS|
================================

> See the :ref:`TF-M PSA template <tfm_psa_template>` for reference implementation of the PSA Attestation API in the |NCS|.

Reviewer: For later: 
Once we have a recommended path for partitioning and such for the nRF54L15, we should update these docs to reflect that.

.. _ug_psa_certified_api_overview_fw_update:

PSA Firmware Update API
***********************

The |NCS| does not implement the PSA Firmware Update API.

Instead, other options are available for the immutable bootloader and the upgradable bootloader.
See :ref:`app_bootloaders` for more information on available bootloaders.

> The bootloaders supported in the |NCS| fulfill requirements by PSA Certified, and several of our devices are already among `PSA Certified Products`_.

Reviewer: I suggest we link to https://www.nordicsemi.com/Products/Technologies/Security instead
